Migrated loss functions to ModuleInfos

mikaylagawarecki opened the pull request:
Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)

#### On test times
This PR increases test time by ~58s locally 
Before this PR:
```
>>> python test/test_nn.py -k Loss
Ran 1003 tests in 28.977s
```
After this PR 
```
>>> python test/test_nn.py -k Loss
Ran 368 tests in 23.073s
```

```
>>> python test/test_modules.py -k Loss
Ran 836 tests in 63.900s
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115584
* #115617

mikaylagawarecki committed:
Migrated loss functions to ModuleInfos

[ghstack-poisoned]

mikaylagawarecki commented on torch/testing/_internal/common_nn.py:
```diff
@@ -2953,7 +2944,7 @@ def _multimarginloss_reference(input, target_idx, p, margin, weight):
     for i in range(0, len(input)):
         if i != target_idx:
-            output += max(0, weight[target_idx] * (margin - input[target_idx] + input[i]) ** p)
+            output += weight[target_idx] * (max(0, (margin - input[target_idx] + input[i])) ** p)
```
This reference function was implemented wrongly, see docs
<img width="701" alt="Screenshot 2023-12-11 at 3 38 01 PM" src="https://github.com/pytorch/pytorch/assets/35276741/2dbbe71f-db4b-4324-92c2-5f299d9ea579">
and the implementation

https://github.com/pytorch/pytorch/blob/585aea6e770f601159e85fa4cc39ce7f67155fdf/aten/src/ATen/native/LossMultiMargin.cpp#L37-L45

mikaylagawarecki commented on test/test_modules.py:
```diff
@@ -304,6 +304,8 @@ def _traverse_obj(self, obj, func):
             return {name: self._traverse_obj(o, func) for name, o in obj.items()}
         elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)):
             return func(obj)
+        elif isinstance(obj, (int, float, bool, complex, str)):
```
Changed this because otherwise it was converting `Tuple[int]` to `Tuple[None]` in the noncontiguous outputs when converting lists of integers in `test_non_contiguous_tensors`

mikaylagawarecki replied:
good point

mikaylagawarecki committed:
Update on "Migrated loss functions to ModuleInfos"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Migrated loss functions to ModuleInfos"

Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Migrated loss functions to ModuleInfos"

Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Migrated loss functions to ModuleInfos"

Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Migrated loss functions to ModuleInfos"

Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)




[ghstack-poisoned]

janeyx99 commented on torch/testing/_internal/common_modules.py:
```diff
@@ -1657,6 +2160,158 @@ def module_inputs_torch_nn_LogSigmoid(module_info, device, dtype, requires_grad,
+            return multilabelmarginloss_reference(i, t, **constructor_kwargs)
+
+        module_inputs.append(
+            ModuleInput(constructor_input=FunctionInput(**constructor_kwargs),
```
There's another MultilabelMarginLoss for 2d where the inputsize is (5, 10). Is that one omitted here cuz redundancy?

mikaylagawarecki replied:
oh good catch, thanks!

janeyx99 commented on torch/testing/_internal/common_modules.py:
```diff
@@ -3103,11 +3831,60 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
+               module_inputs_func=module_inputs_torch_nn_BCELoss,
+               skips=(
+                   # No channels_last support for loss functions.
+                   DecorateInfo(unittest.skip("Skipped!"), 'TestModule', 'test_memory_format'),
```
Is there a way to defacto have these be a part of all loss functions? Like the default skip list has one item already if it's a loss function, and the one item is `DecorateInfo(unittest.skip("Skipped!"), 'TestModule', 'test_memory_format')`?

janeyx99 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3078,918 +3069,7 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
-        default_dtype=torch.double,
-    ),
-]
+criterion_tests = []
```
Is it important to still keep this around? Could we remove references to criterion_tests instead of having the empty list here?

mikaylagawarecki replied:
There are [a couple more tests being added to this list below](https://github.com/pytorch/pytorch/blob/b4defcb742e229e6343f3e21bab8b58d2ac1d8e4/torch/testing/_internal/common_nn.py#L3112-L3192) that weren't migrated in this PR

janeyx99 replied:
ah i see!

janeyx99 reviewed (commented):
Looks good overall--do you have an estimate of how much longer the new tests will take? I think this PR increases test coverage, and it'd be good to know the TTS add

mikaylagawarecki commented on torch/testing/_internal/common_modules.py:
```diff
@@ -349,6 +420,105 @@ def module_inputs_torch_nn_GaussianNLLLoss(module_info, device, dtype, requires_
+        ('reduction_none', {'reduction': 'none'}),
+        ('full', {'full': True}),
+        ('no_log_input', {'log_input': False}),
+        ('full_no_log_input', {'full': True, 'log_input': False}),
```
This is the default that corresponds to the ('', {}) on line 428

mikaylagawarecki commented on torch/testing/_internal/common_modules.py:
```diff
@@ -315,12 +356,42 @@ def reference_fn(m, p, i, t, constructor_kwargs=constructor_kwargs):
-                        forward_input=FunctionInput(make_input((15, 10)),
-                                                    torch.empty(15, device=device).uniform_().mul(10).floor().long()),
+                        forward_input=FunctionInput(make_input((15, 4)),
+                                                    torch.empty(15, device=device).uniform_().mul(4).floor().long()),
```
Wanted to decrease the test time since there are quite a lot of sample inputs, I don't think 10 is a special number here so any other number of classes is probably also good :)

mikaylagawarecki commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3052,16 +3043,16 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
-        return output.sum()
+        output = output.sum()
     output = output.to(dt)
     return output
```
Hm the problem here is that on this line the dtype of `log_probs` is promoted to `float64`, so the `output.to(dt) at the end is necessary for all `reduction` types so that `float32` input --> `float32` output

mikaylagawarecki committed:
Update on "Migrated loss functions to ModuleInfos"

Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)

#### On test times
This PR increases test time by ~58s
Before this PR:
```
>>> python test/test_nn.py -k Loss
Ran 1003 tests in 28.977s
```
After this PR 
```
>>> python test/test_nn.py -k Loss
Ran 368 tests in 23.073s
```

```
>>> python test/test_modules.py -k Loss
Ran 836 tests in 63.900s
```




[ghstack-poisoned]

mikaylagawarecki commented:
@janeyx99 Updated the PR description with the change in test time

mikaylagawarecki committed:
Update on "Migrated loss functions to ModuleInfos"

Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)

#### On test times
This PR increases test time by ~58s locally 
Before this PR:
```
>>> python test/test_nn.py -k Loss
Ran 1003 tests in 28.977s
```
After this PR 
```
>>> python test/test_nn.py -k Loss
Ran 368 tests in 23.073s
```

```
>>> python test/test_modules.py -k Loss
Ran 836 tests in 63.900s
```




[ghstack-poisoned]

janeyx99 reviewed (approved):
LGTM!

mikaylagawarecki commented:
@pytorchbot merge

atalman commented:
@pytorchbot revert -m "OSS CI oncall, breaks slow test" -c nosignal

mikaylagawarecki committed:
Update on "Migrated loss functions to ModuleInfos"

Migrates most tests in `common_nn.py:criterion_tests` to ModuleInfos.

**I can split this up if it is too large to review**

What this PR does not include:
- [`no_batch_dim` tests](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L3995-L4112)
- [tests that use the functional variant of the loss function and `wrap_functional`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_nn.py#L1079-L1128)

#### On test times
This PR increases test time by ~58s locally 
Before this PR:
```
>>> python test/test_nn.py -k Loss
Ran 1003 tests in 28.977s
```
After this PR 
```
>>> python test/test_nn.py -k Loss
Ran 368 tests in 23.073s
```

```
>>> python test/test_modules.py -k Loss
Ran 836 tests in 63.900s
```




[ghstack-poisoned]

janeyx99 commented:
it’s an OOM for MSELoss—is this expected? if so, we should just add one of those memory decorators 

https://hud.pytorch.org/pytorch/pytorch/commit/f138b08d2e9c8d676f2a404e97d773f42132b0c7

mikaylagawarecki commented:
ciflow/slow seems to be passing now

Removed the extremely large ModuleInput with shape (5, 68, 64, 64) in MSELoss for mpw no other moduleinputs have such a big in[it tensor, going to look into where this large input was first added (it seems legacy) and might add it back in a follow up

mikaylagawarecki commented:
@pytorchbot merge
